test/test_autograd.py

```diff
@@ -7003,6 +7013,234 @@ def test_size_check(self):
+            self.assertFalse(t.requires_grad)
+            self.assertEqual(p, foo * 2)
+            self.assertEqual(t, bar * 2)
+
+    def test_view_inplace(self):
+        foo = torch.rand(2)
+        bar = torch.ones(2)
+
+        with fwAD.dual_level():
+            # Note that in this test, we use "update" to mean computing the right tangent for the dual
+            # All the inplace operations here are expected to update the primal value of the Tensors but
+            # not always their tangents.
+            # Also all mentions of "non differentiable view" here means non forward differentiable view
+            # unless specified otherwise.
+
+            # Check that inplace ops do not update non-differentiable views
+            cloned_foo = foo.clone()
+            cloned_bar = bar.clone()
+            # Non differentiable view
+            dual = fwAD.make_dual(cloned_foo, cloned_bar)
+            dual *= 2
+            # Check that non differentiable view was not updated
+            self.assertIsNone(fwAD.unpack_dual(foo)[1])
+            # Check that the computed result is correct
+            self.assertEqual(cloned_bar, bar * 2)
+            self.assertEqual(fwAD.unpack_dual(dual)[1], bar * 2)
+            self.assertEqual(cloned_foo, foo * 2)
+            self.assertEqual(fwAD.unpack_dual(dual)[0], foo * 2)
+            # Other non differentiable view
+            dual_primal, dual_tangent = fwAD.unpack_dual(dual)
+            self.assertIsNone(fwAD.unpack_dual(dual_primal)[1])
+            self.assertIsNone(fwAD.unpack_dual(dual_tangent)[1])
+            dual_primal *= 2
+            # Ensure dual's tangent did not change
+            self.assertEqual(fwAD.unpack_dual(dual)[0], foo * 4)
+            self.assertEqual(fwAD.unpack_dual(dual)[1], bar * 2)
+            dual_tangent *= 2
+            # Ensure dual's primal did not change
+            self.assertEqual(fwAD.unpack_dual(dual)[0], foo * 4)
+            self.assertEqual(fwAD.unpack_dual(dual)[1], bar * 4)
+
+            # Check that inplace ops do update differentiable view but stop at non differentiable ones
+            cloned_foo = foo.clone()
+            cloned_bar = bar.clone()
+            # A non differentiable view
+            dual = fwAD.make_dual(cloned_foo, cloned_bar)
+            # A differentiable view
+            view = dual.narrow(0, 0, 1)
+            view *= 2
+            # Check that non differentiable view was not updated
+            self.assertIsNone(fwAD.unpack_dual(foo)[1])
+            # Check that differentiable view was updated
+            self.assertEqual(fwAD.unpack_dual(dual)[1], torch.tensor([2., 1.]))
+            self.assertEqual(fwAD.unpack_dual(view)[1], torch.tensor([2.]))
+
+            # Check that we track differentiable view even for Tensors that are not dual
+            baz = torch.rand(2)
+            baz += dual
+            self.assertEqual(fwAD.unpack_dual(baz)[1], fwAD.unpack_dual(dual)[1])
```

Comment:
it still really feels like you should have a named tuple return and/or explicit tangent()/primal() accessors

Reply:
I agree, added to the follow up list.